File: src/main/java/org/cubexmc/metro/train/TrainMovementAssistController.java
package org.cubexmc.metro.train;

import org.bukkit.entity.Minecart;

/**
 * Handles safe-mode movement assist for stalled minecarts.
 */
public class TrainMovementAssistController {

    private final TrainSession session;
    private final TrainScheduler trainScheduler;
    private final TrainPhysicsController physicsController;
    private final Runnable cancelHandler;
    private final Runnable passengerExitHandler;
    private Object movementAssistTaskId;

    public TrainMovementAssistController(TrainSession session, TrainScheduler trainScheduler,
            TrainPhysicsController physicsController, Runnable cancelHandler, Runnable passengerExitHandler) {
        this.session = session;
        this.trainScheduler = trainScheduler;
        this.physicsController = physicsController;
        this.cancelHandler = cancelHandler;
        this.passengerExitHandler = passengerExitHandler;
    }

    public void start() {
        stop();
        if (!session.getPlugin().getConfigFacade().isSafeModeMovementAssist()
                || session.getMinecart() == null) {
            return;
        }
        long interval = Math.max(1L, session.getPlugin().getConfigFacade().getSafeModeStallRecoveryTicks());
        movementAssistTaskId = trainScheduler.entityRun(session.getMinecart(), this::recoverStalledMinecart,
                interval, interval);
    }

    public void stop() {
        trainScheduler.cancel(movementAssistTaskId);
        movementAssistTaskId = null;
    }

    private void recoverStalledMinecart() {
        if (!session.getPlugin().getConfigFacade().isSafeModeMovementAssist()) {
            stop();
            return;
        }
        Minecart minecart = session.getMinecart();
        if (minecart == null || minecart.isDead() || !minecart.isValid()) {
            cancelHandler.run();
            return;
        }
        if (!physicsController.canRecoverStalledMinecart(session)) {
            return;
        }
        if (!session.isPassengerStillRiding()) {
            passengerExitHandler.run();
            return;
        }

        double minCruiseSpeed = Math.max(0.01,
                session.getPlugin().getConfigFacade().getSafeModeMinCruiseSpeed());
        if (!physicsController.isBelowCruiseSpeed(minecart, minCruiseSpeed)) {
            return;
        }

        double targetSpeed = physicsController.resolveAssistSpeed(minecart,
                session.getPlugin().getConfigFacade().getCartSpeed(), minCruiseSpeed);
        minecart.setVelocity(physicsController.buildAssistVelocity(session.getLastTravelDirection(), targetSpeed));
    }
}

File: src/main/java/org/cubexmc/metro/train/TrainMovementAssistController.kt
package org.cubexmc.metro.train

import kotlin.math.max

/**
 * Handles safe-mode movement assist for stalled minecarts.
 */
class TrainMovementAssistController(
    private val session: TrainSession,
    private val trainScheduler: TrainScheduler,
    private val physicsController: TrainPhysicsController,
    private val cancelHandler: Runnable,
    private val passengerExitHandler: Runnable,
) {
    private var movementAssistTaskId: Any? = null

    fun start() {
        stop()
        val minecart = session.minecart
        if (!session.plugin.configFacade.isSafeModeMovementAssist() || minecart == null) {
            return
        }
        val interval = max(1L, session.plugin.configFacade.getSafeModeStallRecoveryTicks())
        movementAssistTaskId = trainScheduler.entityRun(
            minecart,
            Runnable { recoverStalledMinecart() },
            interval,
            interval,
        )
    }

    fun stop() {
        trainScheduler.cancel(movementAssistTaskId)
        movementAssistTaskId = null
    }

    private fun recoverStalledMinecart() {
        if (!session.plugin.configFacade.isSafeModeMovementAssist()) {
            stop()
            return
        }
        val minecart = session.minecart
        if (minecart == null || minecart.isDead || !minecart.isValid) {
            cancelHandler.run()
            return
        }
        if (!physicsController.canRecoverStalledMinecart(session)) {
            return
        }
        if (!session.isPassengerStillRiding()) {
            passengerExitHandler.run()
            return
        }

        val minCruiseSpeed = max(0.01, session.plugin.configFacade.getSafeModeMinCruiseSpeed())
        if (!physicsController.isBelowCruiseSpeed(minecart, minCruiseSpeed)) {
            return
        }

        val targetSpeed = physicsController.resolveAssistSpeed(
            minecart,
            session.plugin.configFacade.getCartSpeed(),
            minCruiseSpeed,
        )
        val lastTravelDirection = session.lastTravelDirection ?: return
        minecart.velocity = physicsController.buildAssistVelocity(lastTravelDirection, targetSpeed)
    }
}
